File: packages/sdk-go/metrics.go
package rollgate

import (
	"strconv"
	"strings"
	"sync"
	"sync/atomic"
)

// MetricsSnapshot holds a snapshot of SDK metrics.
type MetricsSnapshot struct {
	// Request metrics
	TotalRequests    int64
	SuccessfulRequests int64
	FailedRequests   int64

	// Latency metrics (in milliseconds)
	AverageLatency float64
	MinLatency     int64
	MaxLatency     int64
	P50Latency     int64
	P95Latency     int64
	P99Latency     int64

	// Cache metrics
	CacheHits      int64
	CacheMisses    int64
	CacheStaleHits int64
	CacheHitRate   float64

	// Circuit breaker metrics
	CircuitState         CircuitState
	CircuitOpenCount     int64
	CircuitHalfOpenCount int64

	// Evaluation metrics
	TotalEvaluations int64
	EvaluationTimeAvgMs float64

	// Error breakdown
	NetworkErrors    int64
	AuthErrors       int64
	RateLimitErrors  int64
	ServerErrors     int64
}

// SDKMetrics collects metrics about SDK operations.
type SDKMetrics struct {
	mu sync.RWMutex

	// Request counters
	totalRequests     int64
	successfulRequests int64
	failedRequests    int64

	// Latency tracking
	latencies []int64

	// Cache metrics
	cacheHits      int64
	cacheMisses    int64
	cacheStaleHits int64

	// Circuit breaker
	circuitState         CircuitState
	circuitOpenCount     int64
	circuitHalfOpenCount int64

	// Evaluations
	totalEvaluations  int64
	evaluationTimeSum int64

	// Errors
	networkErrors   int64
	authErrors      int64
	rateLimitErrors int64
	serverErrors    int64
}

// NewSDKMetrics creates a new SDKMetrics instance.
func NewSDKMetrics() *SDKMetrics {
	return &SDKMetrics{
		latencies:    make([]int64, 0, 1000),
		circuitState: CircuitStateClosed,
	}
}

// RecordRequest records a request with its latency and success status.
func (m *SDKMetrics) RecordRequest(latencyMs int64, success bool, errCategory ErrorCategory) {
	atomic.AddInt64(&m.totalRequests, 1)

	if success {
		atomic.AddInt64(&m.successfulRequests, 1)
	} else {
		atomic.AddInt64(&m.failedRequests, 1)

		switch errCategory {
		case ErrorCategoryNetwork:
			atomic.AddInt64(&m.networkErrors, 1)
		case ErrorCategoryAuth:
			atomic.AddInt64(&m.authErrors, 1)
		case ErrorCategoryRateLimit:
			atomic.AddInt64(&m.rateLimitErrors, 1)
		case ErrorCategoryServer:
			atomic.AddInt64(&m.serverErrors, 1)
		}
	}

	m.mu.Lock()
	m.latencies = append(m.latencies, latencyMs)
	// Keep only last 1000 latencies
	if len(m.latencies) > 1000 {
		m.latencies = m.latencies[len(m.latencies)-1000:]
	}
	m.mu.Unlock()
}

// RecordCacheHit records a cache hit.
func (m *SDKMetrics) RecordCacheHit(stale bool) {
	if stale {
		atomic.AddInt64(&m.cacheStaleHits, 1)
	} else {
		atomic.AddInt64(&m.cacheHits, 1)
	}
}

// RecordCacheMiss records a cache miss.
func (m *SDKMetrics) RecordCacheMiss() {
	atomic.AddInt64(&m.cacheMisses, 1)
}

// RecordCircuitStateChange records a circuit breaker state change.
func (m *SDKMetrics) RecordCircuitStateChange(state CircuitState) {
	m.mu.Lock()
	defer m.mu.Unlock()

	m.circuitState = state
	switch state {
	case CircuitStateOpen:
		m.circuitOpenCount++
	case CircuitStateHalfOpen:
		m.circuitHalfOpenCount++
	}
}

// RecordEvaluation records a flag evaluation with its duration.
func (m *SDKMetrics) RecordEvaluation(durationNs int64) {
	atomic.AddInt64(&m.totalEvaluations, 1)
	atomic.AddInt64(&m.evaluationTimeSum, durationNs/1000000) // Convert to ms
}

// Snapshot returns a snapshot of all metrics.
func (m *SDKMetrics) Snapshot() MetricsSnapshot {
	m.mu.RLock()
	defer m.mu.RUnlock()

	snapshot := MetricsSnapshot{
		TotalRequests:      atomic.LoadInt64(&m.totalRequests),
		SuccessfulRequests: atomic.LoadInt64(&m.successfulRequests),
		FailedRequests:     atomic.LoadInt64(&m.failedRequests),

		CacheHits:      atomic.LoadInt64(&m.cacheHits),
		CacheMisses:    atomic.LoadInt64(&m.cacheMisses),
		CacheStaleHits: atomic.LoadInt64(&m.cacheStaleHits),

		CircuitState:         m.circuitState,
		CircuitOpenCount:     m.circuitOpenCount,
		CircuitHalfOpenCount: m.circuitHalfOpenCount,

		TotalEvaluations: atomic.LoadInt64(&m.totalEvaluations),

		NetworkErrors:   atomic.LoadInt64(&m.networkErrors),
		AuthErrors:      atomic.LoadInt64(&m.authErrors),
		RateLimitErrors: atomic.LoadInt64(&m.rateLimitErrors),
		ServerErrors:    atomic.LoadInt64(&m.serverErrors),
	}

	// Calculate cache hit rate
	totalCacheOps := snapshot.CacheHits + snapshot.CacheMisses + snapshot.CacheStaleHits
	if totalCacheOps > 0 {
		snapshot.CacheHitRate = float64(snapshot.CacheHits+snapshot.CacheStaleHits) / float64(totalCacheOps)
	}

	// Calculate latency stats
	if len(m.latencies) > 0 {
		snapshot.MinLatency, snapshot.MaxLatency, snapshot.AverageLatency = m.calculateLatencyStats()
		snapshot.P50Latency = m.percentile(50)
		snapshot.P95Latency = m.percentile(95)
		snapshot.P99Latency = m.percentile(99)
	}

	// Calculate evaluation time average
	if snapshot.TotalEvaluations > 0 {
		snapshot.EvaluationTimeAvgMs = float64(atomic.LoadInt64(&m.evaluationTimeSum)) / float64(snapshot.TotalEvaluations)
	}

	return snapshot
}

func (m *SDKMetrics) calculateLatencyStats() (min, max int64, avg float64) {
	if len(m.latencies) == 0 {
		return 0, 0, 0
	}

	min = m.latencies[0]
	max = m.latencies[0]
	var sum int64

	for _, l := range m.latencies {
		if l < min {
			min = l
		}
		if l > max {
			max = l
		}
		sum += l
	}

	avg = float64(sum) / float64(len(m.latencies))
	return
}

func (m *SDKMetrics) percentile(p int) int64 {
	if len(m.latencies) == 0 {
		return 0
	}

	// Simple percentile calculation (not perfectly accurate but good enough)
	sorted := make([]int64, len(m.latencies))
	copy(sorted, m.latencies)

	// Simple bubble sort for small arrays
	for i := 0; i < len(sorted)-1; i++ {
		for j := 0; j < len(sorted)-i-1; j++ {
			if sorted[j] > sorted[j+1] {
				sorted[j], sorted[j+1] = sorted[j+1], sorted[j]
			}
		}
	}

	idx := (p * len(sorted)) / 100
	if idx >= len(sorted) {
		idx = len(sorted) - 1
	}

	return sorted[idx]
}

// Reset clears all metrics.
func (m *SDKMetrics) Reset() {
	m.mu.Lock()
	defer m.mu.Unlock()

	atomic.StoreInt64(&m.totalRequests, 0)
	atomic.StoreInt64(&m.successfulRequests, 0)
	atomic.StoreInt64(&m.failedRequests, 0)
	m.latencies = make([]int64, 0, 1000)
	atomic.StoreInt64(&m.cacheHits, 0)
	atomic.StoreInt64(&m.cacheMisses, 0)
	atomic.StoreInt64(&m.cacheStaleHits, 0)
	m.circuitOpenCount = 0
	m.circuitHalfOpenCount = 0
	atomic.StoreInt64(&m.totalEvaluations, 0)
	atomic.StoreInt64(&m.evaluationTimeSum, 0)
	atomic.StoreInt64(&m.networkErrors, 0)
	atomic.StoreInt64(&m.authErrors, 0)
	atomic.StoreInt64(&m.rateLimitErrors, 0)
	atomic.StoreInt64(&m.serverErrors, 0)
}

// ToPrometheus exports metrics in Prometheus text format.
func (m *SDKMetrics) ToPrometheus(prefix string) string {
	snap := m.Snapshot()
	var b strings.Builder

	metric := func(name string, value interface{}, help, mtype string) {
		fullName := prefix + "_" + name
		if help != "" {
			b.WriteString("# HELP ")
			b.WriteString(fullName)
			b.WriteString(" ")
			b.WriteString(help)
			b.WriteString("\n")
		}
		if mtype != "" {
			b.WriteString("# TYPE ")
			b.WriteString(fullName)
			b.WriteString(" ")
			b.WriteString(mtype)
			b.WriteString("\n")
		}
		b.WriteString(fullName)
		b.WriteString(" ")
		switch v := value.(type) {
		case int64:
			b.WriteString(strconv.FormatInt(v, 10))
		case float64:
			b.WriteString(strconv.FormatFloat(v, 'f', 2, 64))
		}
		b.WriteString("\n")
	}

	// Request metrics
	metric("requests_total", snap.TotalRequests, "Total number of requests", "counter")
	metric("requests_success_total", snap.SuccessfulRequests, "Total successful requests", "counter")
	metric("requests_failed_total", snap.FailedRequests, "Total failed requests", "counter")

	// Latency metrics
	metric("latency_avg_ms", snap.AverageLatency, "Average request latency in milliseconds", "gauge")
	metric("latency_p50_ms", snap.P50Latency, "50th percentile latency", "gauge")
	metric("latency_p95_ms", snap.P95Latency, "95th percentile latency", "gauge")
	metric("latency_p99_ms", snap.P99Latency, "99th percentile latency", "gauge")

	// Cache metrics
	metric("cache_hits_total", snap.CacheHits, "Total cache hits", "counter")
	metric("cache_misses_total", snap.CacheMisses, "Total cache misses", "counter")
	metric("cache_hit_rate", snap.CacheHitRate, "Cache hit rate percentage", "gauge")

	// Circuit breaker metrics
	metric("circuit_opens_total", snap.CircuitOpenCount, "Total circuit breaker opens", "counter")
	var circuitValue float64
	switch snap.CircuitState {
	case CircuitStateClosed:
		circuitValue = 0
	case CircuitStateOpen:
		circuitValue = 1
	case CircuitStateHalfOpen:
		circuitValue = 0.5
	}
	metric("circuit_state", circuitValue, "Circuit breaker state (0=closed, 0.5=half-open, 1=open)", "gauge")

	// Evaluation metrics
	metric("evaluations_total", snap.TotalEvaluations, "Total flag evaluations", "counter")
	metric("evaluation_avg_time_ms", snap.EvaluationTimeAvgMs, "Average evaluation time in milliseconds", "gauge")

	// Error metrics
	metric("errors_network_total", snap.NetworkErrors, "Total network errors", "counter")
	metric("errors_auth_total", snap.AuthErrors, "Total authentication errors", "counter")
	metric("errors_ratelimit_total", snap.RateLimitErrors, "Total rate limit errors", "counter")
	metric("errors_server_total", snap.ServerErrors, "Total server errors", "counter")

	return b.String()
}

File: packages/sdk-java/src/main/java/io/rollgate/Metrics.java
package io.rollgate;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicLong;
import java.util.concurrent.atomic.AtomicReference;

/**
 * SDK metrics collection for monitoring and observability.
 */
public class Metrics {

    private final AtomicLong totalRequests = new AtomicLong(0);
    private final AtomicLong successfulRequests = new AtomicLong(0);
    private final AtomicLong failedRequests = new AtomicLong(0);
    private final AtomicLong cacheHits = new AtomicLong(0);
    private final AtomicLong cacheMisses = new AtomicLong(0);
    private final AtomicLong circuitOpens = new AtomicLong(0);
    private final AtomicLong circuitCloses = new AtomicLong(0);
    private final AtomicLong totalEvaluations = new AtomicLong(0);

    private final ConcurrentHashMap<String, AtomicLong> errorsByCategory = new ConcurrentHashMap<>();
    private final ConcurrentHashMap<String, FlagStats> flagStats = new ConcurrentHashMap<>();

    private final List<Long> latencies = Collections.synchronizedList(new ArrayList<>());
    private static final int MAX_LATENCY_HISTORY = 1000;

    private final AtomicReference<CircuitBreaker.State> circuitState =
        new AtomicReference<>(CircuitBreaker.State.CLOSED);

    private final long startTime = System.currentTimeMillis();
    private final AtomicLong lastRequestAt = new AtomicLong(0);

    /**
     * Flag evaluation statistics.
     */
    public static class FlagStats {
        private final AtomicLong evaluations = new AtomicLong(0);
        private final AtomicLong trueCount = new AtomicLong(0);
        private final AtomicLong totalTimeMs = new AtomicLong(0);

        public void record(boolean result, long evaluationTimeMs) {
            evaluations.incrementAndGet();
            if (result) {
                trueCount.incrementAndGet();
            }
            totalTimeMs.addAndGet(evaluationTimeMs);
        }

        public long getEvaluations() { return evaluations.get(); }
        public long getTrueCount() { return trueCount.get(); }
        public long getFalseCount() { return evaluations.get() - trueCount.get(); }
        public double getTrueRate() {
            long total = evaluations.get();
            return total > 0 ? (double) trueCount.get() / total * 100 : 0;
        }
        public double getAvgEvaluationTimeMs() {
            long total = evaluations.get();
            return total > 0 ? (double) totalTimeMs.get() / total : 0;
        }
    }

    /**
     * Metrics snapshot.
     */
    public static class Snapshot {
        public final long totalRequests;
        public final long successfulRequests;
        public final long failedRequests;
        public final double successRate;
        public final double errorRate;

        public final double avgLatencyMs;
        public final long minLatencyMs;
        public final long maxLatencyMs;
        public final long p50LatencyMs;
        public final long p95LatencyMs;
        public final long p99LatencyMs;

        public final long cacheHits;
        public final long cacheMisses;
        public final double cacheHitRate;

        public final Map<String, Long> errorsByCategory;

        public final long circuitOpens;
        public final long circuitCloses;
        public final CircuitBreaker.State circuitState;

        public final long totalEvaluations;
        public final Map<String, FlagStats> flagStats;

        public final long uptimeMs;
        public final long lastRequestAt;

        Snapshot(Metrics m) {
            this.totalRequests = m.totalRequests.get();
            this.successfulRequests = m.successfulRequests.get();
            this.failedRequests = m.failedRequests.get();
            this.successRate = totalRequests > 0 ? (double) successfulRequests / totalRequests * 100 : 0;
            this.errorRate = totalRequests > 0 ? (double) failedRequests / totalRequests * 100 : 0;

            List<Long> sortedLatencies;
            synchronized (m.latencies) {
                sortedLatencies = new ArrayList<>(m.latencies);
            }
            Collections.sort(sortedLatencies);

            this.avgLatencyMs = calculateAverage(sortedLatencies);
            this.minLatencyMs = sortedLatencies.isEmpty() ? 0 : sortedLatencies.get(0);
            this.maxLatencyMs = sortedLatencies.isEmpty() ? 0 : sortedLatencies.get(sortedLatencies.size() - 1);
            this.p50LatencyMs = calculatePercentile(sortedLatencies, 50);
            this.p95LatencyMs = calculatePercentile(sortedLatencies, 95);
            this.p99LatencyMs = calculatePercentile(sortedLatencies, 99);

            this.cacheHits = m.cacheHits.get();
            this.cacheMisses = m.cacheMisses.get();
            long totalCache = cacheHits + cacheMisses;
            this.cacheHitRate = totalCache > 0 ? (double) cacheHits / totalCache * 100 : 0;

            this.errorsByCategory = new HashMap<>();
            m.errorsByCategory.forEach((k, v) -> this.errorsByCategory.put(k, v.get()));

            this.circuitOpens = m.circuitOpens.get();
            this.circuitCloses = m.circuitCloses.get();
            this.circuitState = m.circuitState.get();

            this.totalEvaluations = m.totalEvaluations.get();
            this.flagStats = new HashMap<>(m.flagStats);

            this.uptimeMs = System.currentTimeMillis() - m.startTime;
            this.lastRequestAt = m.lastRequestAt.get();
        }

        private static double calculateAverage(List<Long> values) {
            if (values.isEmpty()) return 0;
            return values.stream().mapToLong(Long::longValue).average().orElse(0);
        }

        private static long calculatePercentile(List<Long> sortedValues, int percentile) {
            if (sortedValues.isEmpty()) return 0;
            int index = (int) Math.ceil((percentile / 100.0) * sortedValues.size()) - 1;
            return sortedValues.get(Math.max(0, index));
        }
    }

    /**
     * Record a completed request.
     */
    public void recordRequest(int statusCode, long latencyMs, boolean cacheHit, String errorCategory) {
        totalRequests.incrementAndGet();
        lastRequestAt.set(System.currentTimeMillis());

        boolean success = statusCode >= 200 && statusCode < 400;
        if (success) {
            successfulRequests.incrementAndGet();
        } else {
            failedRequests.incrementAndGet();
        }

        if (cacheHit) {
            cacheHits.incrementAndGet();
        } else {
            cacheMisses.incrementAndGet();
        }

        synchronized (latencies) {
            latencies.add(latencyMs);
            while (latencies.size() > MAX_LATENCY_HISTORY) {
                latencies.remove(0);
            }
        }

        if (errorCategory != null && !errorCategory.isEmpty()) {
            errorsByCategory.computeIfAbsent(errorCategory, k -> new AtomicLong(0)).incrementAndGet();
        }
    }

    /**
     * Record a flag evaluation.
     */
    public void recordEvaluation(String flagKey, boolean result, long evaluationTimeMs) {
        totalEvaluations.incrementAndGet();
        flagStats.computeIfAbsent(flagKey, k -> new FlagStats()).record(result, evaluationTimeMs);
    }

    /**
     * Record circuit breaker state change.
     */
    public void recordCircuitStateChange(CircuitBreaker.State newState) {
        CircuitBreaker.State oldState = circuitState.getAndSet(newState);
        if (newState == CircuitBreaker.State.OPEN && oldState != CircuitBreaker.State.OPEN) {
            circuitOpens.incrementAndGet();
        } else if (newState == CircuitBreaker.State.CLOSED && oldState != CircuitBreaker.State.CLOSED) {
            circuitCloses.incrementAndGet();
        }
    }

    /**
     * Get a snapshot of all metrics.
     */
    public Snapshot snapshot() {
        return new Snapshot(this);
    }

    /**
     * Export metrics in Prometheus format.
     */
    public String toPrometheus(String prefix) {
        if (prefix == null) prefix = "rollgate_sdk";

        Snapshot snap = snapshot();
        StringBuilder sb = new StringBuilder();

        appendMetric(sb, prefix, "requests_total", snap.totalRequests, "Total number of requests", "counter");
        appendMetric(sb, prefix, "requests_success_total", snap.successfulRequests, "Total successful requests", "counter");
        appendMetric(sb, prefix, "requests_failed_total", snap.failedRequests, "Total failed requests", "counter");

        appendMetric(sb, prefix, "latency_avg_ms", snap.avgLatencyMs, "Average request latency in milliseconds", "gauge");
        appendMetric(sb, prefix, "latency_p50_ms", snap.p50LatencyMs, "50th percentile latency", "gauge");
        appendMetric(sb, prefix, "latency_p95_ms", snap.p95LatencyMs, "95th percentile latency", "gauge");
        appendMetric(sb, prefix, "latency_p99_ms", snap.p99LatencyMs, "99th percentile latency", "gauge");

        appendMetric(sb, prefix, "cache_hits_total", snap.cacheHits, "Total cache hits", "counter");
        appendMetric(sb, prefix, "cache_misses_total", snap.cacheMisses, "Total cache misses", "counter");
        appendMetric(sb, prefix, "cache_hit_rate", snap.cacheHitRate, "Cache hit rate percentage", "gauge");

        appendMetric(sb, prefix, "circuit_opens_total", snap.circuitOpens, "Total circuit breaker opens", "counter");
        double circuitStateValue = snap.circuitState == CircuitBreaker.State.CLOSED ? 0 :
                                   snap.circuitState == CircuitBreaker.State.OPEN ? 1 : 0.5;
        appendMetric(sb, prefix, "circuit_state", circuitStateValue, "Circuit breaker state (0=closed, 0.5=half-open, 1=open)", "gauge");

        appendMetric(sb, prefix, "evaluations_total", snap.totalEvaluations, "Total flag evaluations", "counter");
        appendMetric(sb, prefix, "uptime_seconds", snap.uptimeMs / 1000.0, "SDK uptime in seconds", "gauge");

        return sb.toString();
    }

    private void appendMetric(StringBuilder sb, String prefix, String name, double value, String help, String type) {
        String fullName = prefix + "_" + name;
        sb.append("# HELP ").append(fullName).append(" ").append(help).append("\n");
        sb.append("# TYPE ").append(fullName).append(" ").append(type).append("\n");
        sb.append(fullName).append(" ").append(value).append("\n");
    }

    /**
     * Reset all metrics.
     */
    public void reset() {
        totalRequests.set(0);
        successfulRequests.set(0);
        failedRequests.set(0);
        cacheHits.set(0);
        cacheMisses.set(0);
        circuitOpens.set(0);
        circuitCloses.set(0);
        totalEvaluations.set(0);
        errorsByCategory.clear();
        flagStats.clear();
        synchronized (latencies) {
            latencies.clear();
        }
        circuitState.set(CircuitBreaker.State.CLOSED);
        lastRequestAt.set(0);
    }
}
